spock-specs/src/test/groovy/org/spockframework/smoke/extension/RepeatableLocalExtensionsSpec.groovy

```diff
@@ -0,0 +1,240 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.spockframework.smoke.extension
+
+import org.spockframework.EmbeddedSpecification
+import org.spockframework.runtime.extension.ExtensionAnnotation
+import org.spockframework.runtime.extension.IAnnotationDrivenExtension
+import org.spockframework.runtime.model.FeatureInfo
+import org.spockframework.runtime.model.FieldInfo
+import org.spockframework.runtime.model.MethodInfo
+import org.spockframework.runtime.model.SpecInfo
+import spock.util.EmbeddedSpecRunner
+import spock.util.mop.ConfineMetaClassChanges
+
+import java.lang.annotation.Repeatable
+import java.lang.annotation.Retention
+import java.lang.annotation.RetentionPolicy
+
+@ConfineMetaClassChanges(EmbeddedSpecRunner)
+class RepeatableLocalExtensionsSpec extends EmbeddedSpecification {
+  def setup() {
+    runner.addClassImport(Foo)
+    runner.addClassImport(FooContainer)
+
+    runner.metaClass.runSpec = {
+      runner.runWithImports """
+class FooSpec extends Specification {
+  static specAnnotationsVisited = 0
+  static featureAnnotationsVisited = 0
+  static fixtureAnnotationsVisited = 0
+  static fieldAnnotationsVisited = 0
+
+  $it
+}
+"""
+    }
+  }
+
+  def "should work properly on spec"() {
+    expect:
+    runner.runWithImports """
+${buildContainer(contained, containerWithList)}
+${'@Foo\n' * individual}
+class FooSpec extends Specification {
+  static specAnnotationsVisited = 0
+  static featureAnnotationsVisited = 0
+  static fixtureAnnotationsVisited = 0
+  static fieldAnnotationsVisited = 0
+
+  def foo() {
+    expect:
+    verifyAll {
+      specAnnotationsVisited == $expected
+      featureAnnotationsVisited == 0
+      fixtureAnnotationsVisited == 0
+      fieldAnnotationsVisited == 0
+    }
+  }
+}
+"""
+
+    where:
+    individual | contained | containerWithList || expected
+    0          | -1        | false             || 0
+    1          | -1        | false             || 1
+    2          | -1        | false             || 2
+    0          | 0         | false             || 0
+    0          | 0         | true              || 0
+    1          | 0         | false             || 1
+    1          | 0         | true              || 1
+    0          | 1         | false             || 1
+    0          | 1         | true              || 1
+    1          | 1         | false             || 2
+    1          | 1         | true              || 2
+    0          | 2         | true              || 2
+    1          | 2         | true              || 3
+  }
+
+  def "should work properly on feature"() {
+    expect:
+    runner.runSpec """
+${buildContainer(contained, containerWithList)}
```

Comment:
How do you suggest that I improve it?
Being very verbose and long in the `where` block having the annotations there manually written and hard to read properly?
Adding some comments to what the test does?
I tried to write easy to understand and read and at the same time concise tests.
The where table specifies how many individual annotations, how many contained in the container annotation manually and if one whether directly or inside a list to use and how many extension calls are expected.

Reply:
I would think about generating the tests based on the code you already have and commit both. I think this where DAMP is more important than DRY [explanation](https://stackoverflow.com/questions/6453235/what-does-damp-not-dry-mean-when-talking-about-unit-tests)

What are your thoughts @szpak @masooh ?

Reply:
I don't think committing both makes too much sense, because the next person will change the generated code not even being aware of the source, the next person after that might again be aware of the source, change the source and re-generate, overwriting the first other persons changes and so on.

I though the numbers actually **are** DAMP.
For example
```
individual | contained | containerWithList || expected
1          | 2         | true              || 3
```
means 1 individual annotation without container, two annotations in the container and the container value as list should result in 3 times the extension executed.
I actually inteded this to be actually good readable. :-(

Reply:
For me the test is very hard to read and it took me a while to find out what happens. 

Nevertheless I must admit that I don't have a better solution for this scenario, I have to think about it. 

Javadoc for buildContainer (some samples what could be generated) and for RepeatableLocalExtensionsSpec (that the test uses a dummy extension to check the behaviour) would help for sure.

Reply:
I added a bit of JavaDoc to hopefully make the tests clearer and easier to interpret :-)